### PR TITLE
Feature/post tweet ui

### PR DESCRIPTION
**Pull request by lee-benjamin**

### CODE REVIEW

File: src/js/components.js [hunk 1]
```diff
@@ -1,85 +1,164 @@
...
+    this.focusedComponentListener = this.focusedComponentListener.bind(this);
   }
 
-  function failureCallback() {
-    ReactDOM.render(e(ReactContainer, {isServerError: true}),
-      document.getElementById("root"));
+  focusedComponentListener(componentName) {
+    let focusedComponent = null;
+    switch(componentName) {
+      case "User Timeline":
```
Comment (kev-chan) [inline src/js/components.js]: create a java like enum in javascript

File: src/js/components.js [hunk 2]
```diff
@@ -1,85 +1,164 @@
...
-      isServerError: props.isServerError
-    };
+    this.state = {tweetText:"", successMessage: ""};
+    this.onChange = this.onChange.bind(this);
+    this.onClick = this.onClick.bind(this);
+    this.getMessageClass = this.getMessageClass.bind(this);
+  }
+
+  onClick() {
+    twitter.postTweet(this.state.tweetText)
```
Comment (kev-chan) [inline src/js/components.js]: let's improve the readability of the success message handling

File: src/js/components.js [hunk 3]
```diff
@@ -1,85 +1,164 @@
...
-      ),
-      e(UserTimeline,
-        {
-          className: "UserTimeline",
-          isServerError: this.state.isServerError,
-          tweets: this.state.userTweets
-        }
+      {className: "PostTweet"},
+      e("h1", {className: "Header"}, "Post a tweet"),
+      e("div", {className: "tweetBoxCharCount"},
```
Comment (kev-chan) [inline src/js/components.js]: let's come up with a better name than tweetBoxCharCount as it encompasses more than just tweetbox and char count

Comment (lee-benjamin) [reply src/js/components.js]: removing this whole div since it's not needed anymore

File: src/js/components.js (hunk 1 above)
Comment (lee-benjamin) [reply src/js/components.js]: done

File: src/js/components.js [hunk 4]
```diff
@@ -205,29 +346,31 @@ class UserTimeline extends React.Component {
...
           className: "getTimelineButton",
-          onClick: this.handleClick,
+          onClick: this.getTweets,
         },
         "Get User Timeline"
       ),
-      e(component, {tweets: this.state.tweets})
+      (this.state.isServerError ?
+        e(ServerError, null) :
+        (this.state.hasNoTweets ? e(NoTweets, null) : Timeline({tweets: this.state.tweets}))
```
Comment (kev-chan) [inline src/js/components.js]: you should be able to go without the hasNoTweets state variable

File: src/js/components.js [hunk 5]
```diff
@@ -1,139 +1,274 @@
...
+
+  onClick() {
+    this.props.changeFocusedTab(this.props.tabName);
   }
 
   render() {
+    return e(
+      "button",
+      {
+        className: "Tab " + (this.props.focusedTab == this.props.tabName ? "focused" : ""),
```
Comment (kev-chan) [inline src/js/components.js]: better to keep this as "Tab" and then add " focused"

File: src/js/components.js [hunk 6]
```diff
@@ -1,139 +1,274 @@
...
-    if (tweets.length == 0) {
-      this.setState({hasNoTweets: true});
-    }
-    else {
-      this.setState({
+    this.setState(
+      {
         isServerError: false,
-        hasNoTweets: false,
+        hasNoTweets: (tweets.length == 0 ? true : false),
```
Comment (kev-chan) [inline src/js/components.js]: better to use the following for defensive coding
tweets && tweets.length ? false : true

File: src/js/components.js [hunk 7]
```diff
@@ -147,7 +282,7 @@ class SearchComponent extends React.Component {
   }
 
   onClick() {
-    filterHomeTimeline(this.state.filterQuery)
+    twitter.filterHomeTimeline(this.state.filterQuery)
```
Comment (kev-chan) [inline src/js/components.js]: you aren't handling the error scenario

File: src/js/components.js (hunk 6 above)
Comment (kev-chan) [reply src/js/components.js]: you should use this method for variable accessing to ensure NPEs don't occur

File: src/js/services/twitter.js [hunk 8]
```diff
@@ -18,6 +18,29 @@ export const getTwitterLink = (tweet) => {
   return urlPrefix + tweet.id;
 }
 
+export const postTweet = (tweet) => {
+  return new Promise((resolve, reject) => {
+    const url = "http://localhost:8080/api/1.0/twitter/tweet";
+    const obj = {
```
Comment (kev-chan) [inline src/js/services/twitter.js]: params is better than obj

File: src/styles/sass/style.scss [hunk 9]
```diff
@@ -27,13 +33,83 @@ $margin: 10px;
...
+    font-size: small;
+    color: gray;
+    bottom: 6px;
+  }
+
+  .PostTweetButton {
+    float: right;
+  }
+
+  .resultMessage {
```
Comment (kev-chan) [inline src/styles/sass/style.scss]: you should clear the float on resultMessage

Comment (lee-benjamin) [reply src/styles/sass/style.scss]: done

File: src/js/services/twitter.js (hunk 8 above)
Comment (lee-benjamin) [reply src/js/services/twitter.js]: done

File: src/js/components.js (hunk 6 above)
Comment (lee-benjamin) [reply src/js/components.js]: done, this is also a neat trick

File: src/js/components.js [hunk 10]
```diff
@@ -1,157 +1,281 @@
...
         {
-          isServerError: this.state.isServerError,
-          tweets: this.state.homeTweets
-        }
+          onChange: this.onChange,
+          maxLength: maxTweetLength
+        },
       ),
-      e(UserTimeline,
+      e("div", {className:"CharCount"}, this.state.tweetText.length),
```
Comment (kev-chan) [inline src/js/components.js]: NPE check here

File: src/js/components.js [hunk 11]
```diff
@@ -1,157 +1,281 @@
...
+      e(
+        "button",
         {
-          className: "UserTimeline",
-          isServerError: this.state.isServerError,
-          tweets: this.state.userTweets
-        }
+          onClick: this.onClick,
+          disabled: !this.state.tweetText.length,
```
Comment (kev-chan) [inline src/js/components.js]: here as well

File: src/js/components.js [hunk 12]
```diff
@@ -278,8 +406,8 @@ function TweetContent(props) {
   return e(
     "div",
     {className: "TweetContent"},
-    e("div", {className: "timeDiv"}, formatDate(props.tweet.createdAt)),
-    e("a", {className: "tweetText", href: getTwitterLink(props.tweet), target: "_blank"},
+    e("div", {className: "timeDiv"}, twitter.formatDate(props.tweet.createdAt)),
```
Comment (kev-chan) [inline src/js/components.js]: NPE check

**Pull request merged**